client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java

```diff
@@ -201,39 +202,21 @@ private List<ShuffleBlockInfo> insertIntoBuffer(
     // this may trigger current WriteBufferManager spill method, which will
     // make the current write buffer discard. So we have to recheck the buffer existence.
     boolean hasRequested = false;
-    if (buffers.containsKey(partitionId)) {
-      WriterBuffer wb = buffers.get(partitionId);
+    WriterBuffer wb = buffers.get(partitionId);
+    if (wb != null) {
       if (wb.askForMemory(serializedDataLength)) {
         requestMemory(required);
         hasRequested = true;
       }
     }
 
-    if (buffers.containsKey(partitionId)) {
+    // The clear operation which triggered by spill operation will cause thread-safe problems
+    wb = buffers.remove(partitionId);
```

Comment:
There seems to be something wrong with the original logic. Such as the following scene.

`buffers.get()[thread-clear] ->  buffers.get()[thread-insert] -> 
wb.getData()[thread-clear] -> wb.addRecord()[thread-insert] -> wb.getMemoryUsed()[thread-clear] (maybe cause free more memory and miss record)`

`buffers.get()[thread-clear] ->  buffers.get()[thread-insert] -> 
wb.getData()[thread-clear] -> wb.getMemoryUsed()[thread-clear] -> wb.addRecord()[thread-insert] (maybe cause miss record)`

PTAL @zuston @jerqi

Reply:
Could you help add the test case to verify this? I'm trying to understand this.

Reply:
The test case is successful in current master branch, so I'm still not catch your thought about this patch.

Reply:
> The test case is successful in current master branch, so I'm still not catch your thought about this patch.

Please repeat this test for 100 times in current master branch. @zuston

Reply:
<img width="1410" alt="Snipaste_2024-01-30_16-17-14" src="https://github.com/apache/incubator-uniffle/assets/11752250/1598bbdd-c350-4d62-a245-1fd232ae4d54">

Reply:
It is my fault.